sound/soc/sof/ipc4-topology.c

```diff
@@ -2106,10 +2111,27 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 	struct sof_ipc4_fw_module *sink_fw_module = sink_widget->module_info;
 	struct sof_ipc4_pipeline *src_pipeline = src_pipe_widget->private;
 	struct sof_ipc4_pipeline *sink_pipeline = sink_pipe_widget->private;
+	u32 core_id = src_pipeline->core_id;
 	struct sof_ipc4_msg msg = {{ 0 }};
 	u32 header, extension;
 	int ret;
 
+	if (sink_pipeline->core_id != core_id) {
+		dev_err(sdev->dev,
+			"pipelines must be on the same core: src: %u (%s), sink: %u (%s)\n",
+			core_id, src_pipe_widget->widget->name,
+			sink_pipeline->core_id, sink_pipe_widget->widget->name);
+		return -EINVAL;
+	}
+
+	if (src_widget->core != core_id || sink_widget->core != core_id) {
+		dev_err(sdev->dev,
+			"widgets must be on core %u: src: %u (%s), sink: %u (%s)\n",
+			core_id, src_widget->core, src_widget->widget->name,
+			sink_widget->core, sink_widget->widget->name);
+		return -EINVAL;
+	}
```

Comment:
same on the source side.

Reply:
right, so no limitation on the connected pipelines and widgets.
The only limitation is that a widget's core must match with the pipeline's core that it is part of?

Reply:
iow, this is not valid:
```c
if (src_pipeline->core_id != src_widget->core) {
...
}

if (sink_pipeline->core_id != sink_widget->core) {
...
}

```

Reply:
> The only limitation is that a widget's core must match with the pipeline's core that it is part of?

Even that part is not clear. When we have DP modules, we will be able to have chained modules on the same pipeline use different cores. So I don't know what the pipeline core would mean, it's an irrelevant notion when DP is used.

Reply:
right, so if all combinations are fine then what to check?

Reply:
and most importantly, how the DP is going to be used? Is it going to have some mark in tplg?
Without DP we do have limitation, no? We should have the limitation checked now (we don't have DP support afaik) and make it conditional when DP is available?

Reply:
IIRC there is nothing advertized anywhere in kernel-accessible manifests or topology that a module is LL or DP, so it could even be the case that only firmware can check for invalid configurations. @lgirdwood can we have the firmware folks confirm what errors cases we could/need to handled in the driver?

Reply:
fwiw, if you send the `CREATE_PIPELINE` message with core_id=1, the firmware will reply with "invalid core id"